### ISSUE TITLE
Add support for Vary response header

### ISSUE THREAD

**Issue by JWCook** [requests-cache/requests-cache#399]
Some servers send `Vary` response headers to inform HTTP caches which request headers to match against. This would be doable, but first it would be worth researching a bit more to see how much additional benefit this would provide compared to #387

Reference: https://datatracker.ietf.org/doc/html/rfc7234#section-4.1

Feedback would be appreciated! Does anyone else have thoughts on this?

**Comment by mjpieters** [comment]
Vary is essential when sending *content negotiation requests*; e.g. `Accept-Encoding: gzip, br`, resulting in a `Vary: Accept-Encoding, Content-Encoding: *gzip or br*` response. That response can not be used to satisfy `Accept-Encoding: deflate` requests, even though every other cache header on the response states it is otherwise still fresh.

While within a single project I don't see a direct use case for `Accept-Encoding`, I would expect `Accept-Language` variations to crop up more readily. You want to treat `Accept-Language: en` differently from `Accept-Language: fr` but only if the server indicates (with `Vary: Accept-Language`) that the response actually differs based on that request header. A request for, say, a binary data file from the same server that doesn't include a `Vary` header can be used regardless of the request `Accept-Language` header.

In other words: Vary helps keep cache bloat down by only requiring additional cache storage for responses that actually vary. Supporting Vary would very much be my preferred approach, over #387.

**Comment by JWCook** [comment]
Thank you for the feedback! I agree that this would be a better solution than #387. That would have been more of an easy-to-implement intermediate solution, basically just a default setting for the existing [`match_headers`](https://requests-cache.readthedocs.io/en/stable/user_guide/matching.html#matching-request-headers) option.

The main thing I'm unsure about is whether supporting `Vary` provides enough value to this library to justify the fair amount of complexity it would add. Do you have a current use case with requests-cache that would benefit from this?

**Comment by mjpieters** [comment]
I personally don't have a use case right now. I just happened to have closed a CDN project last month, so Vary is very fresh in my mind still.

(and if the Varnish source code was easier to read, I'd point you to their implementation of handling Vary. Alas).

**Comment by JWCook** [comment]
Fair enough. Either way, I appreciate the input!

**Comment by allanlei** [comment]
I would also like to put in my support of Vary over #387 

* Our production use case of requests-cache is in a server-to-server environment
  * oEmbed/opengraph unfurling
  * HTML/JSON fetching
* IMO, normalizing the values of headers (like Accept, Accept-Encoding, etc) should be left to the user of requests-cache rather than supported in the lib
  * There's quite a few headers would need to re-invent the normalization like [DPR](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/DPR)

I'm currently checking if Vary is supported before implementing ourselves, so I'm 👍 on this

**Comment by JWCook** [comment]
Thanks for the feedback. If I don't need to worry about normalizing header values, that does simplify things a bit. 

> Maybe a `request_key_fn` and a `response_key_fn`

I believe your own normalization could be done with a single key function, since it [already receives a `match_headers` list](https://github.com/requests-cache/requests-cache/blob/main/requests_cache/cache_keys.py#L34-L40). For the secondary cache key, unless there are other complications I'm not thinking of, I would just call `create_key()` again with the headers specified by `Vary`. For example, if you wanted to normalize `Accept`, it could potentially look something like this:

```python
from requests_cache import CachedSession, create_key

def custom_create_key(request, match_headers, **kwargs):
    # match_headers will be populated by Vary, if we're creating a secondary cache key
    # For this example, if Vary contains Accept, I only care whether or not 'Accept' contains 'gzip'
    if 'Accept' in match_headers and 'Accept' in request.headers:
        request.headers['Accept'] = 'gzip' if 'gzip' in request.headers['Accept'] else None

    # Use base implementation for everything else
    return create_key(request, match_headers=match_headers, **kwargs)

session = CachedSession(key_fn=custom_create_key)
```

**Comment by mjpieters** [comment]
Regarding normalisation: I also would recommend leaving that up to the application. Some applications will care about `Accept-Encoding` weightings (`deflate, gzip;q=1.0, *;q=0.5`) and some won't; e.g. the Varnish reverse proxy cache, in the default configuration, 'normalises' `Accept-Encoding` to *only contain `gzip`*, regardless of whatever the incoming request says.

Bottom line: applications that care about Vary handling will also care about specific normalisations per specific header. Some kind of callback API shoulf suffice here.

**Comment by JWCook** [comment]
Sounds good.

> Some kind of callback API should suffice here.

I think I can adapt the existing `key_fn` callback to work here.

Another detail I wasn't aware of is that unlike CDNs, [browser caches typically don't support multiple variations per URL](https://www.smashingmagazine.com/2017/11/understanding-vary-header/#introducing-vary-as-a-validator). I'd like to do the same, since I'm aiming more for browser cache-like behavior, and I don't think anyone is using requests-cache as a proxy cache (although I could always be proven wrong!).

Normalization and multiple URL variations were the most complex parts I was concerned about adding. Now this looks like a higher value-to-effort ratio than I previously thought, so I will plan on adding Vary support for the 1.0 release.

**Comment by JWCook** [comment]
Some initial changes to support this are in `main`.

**Comment by JWCook** [comment]
There's still some room for improvement, but support for `Vary` is working now in the latest pre-release build.

Give it a try when you get the chance, and let me know if it behaves as you would expect.
